src/Cryptol/Eval/FFI.hs

```diff
@@ -0,0 +1,55 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE LambdaCase     #-}
+
+module Cryptol.Eval.FFI
+  ( addForeignDecls
+  ) where
+
+import           Data.Foldable
+import           Data.IORef
+
+import           Cryptol.Backend.Concrete
+import           Cryptol.Backend.FFI
+import           Cryptol.Backend.Monad
+import           Cryptol.Backend.WordValue
+import           Cryptol.Eval
+import           Cryptol.Eval.Env
+import           Cryptol.Eval.Prims
+import           Cryptol.Eval.Value
+import           Cryptol.ModuleSystem.Env
+import           Cryptol.ModuleSystem.Name
+import           Cryptol.TypeCheck.AST
+import           Cryptol.Utils.Ident
+import           Cryptol.Utils.Panic
+
+addForeignDecls :: ModulePath -> Module ->
+  EvalEnv -> Eval (Maybe ForeignSrc, EvalEnv)
+addForeignDecls path m env = io do
+  foreignSrc <- newIORef Nothing
+  let addForeignDeclGroup e (Recursive ds)   = foldlM addForeignDecl e ds
```

Comment:
Probably should just refuse to consider recursive groups containing foreign declarations, as this shouldn't happen.

Reply:
Yeah, I was thinking that it could happen when we add the ability to provide backup Cryptol implementations of foreign functions, but I can remove it for now.